distributed/client.py

```diff
@@ -1294,34 +1297,65 @@ async def _update_scheduler_info(self):
         except OSError:
             logger.debug("Not able to query scheduler for identity")
 
-    async def _wait_for_workers(self, n_workers=0, timeout=None):
+    async def _wait_for_workers(
+        self,
+        n_workers: int = 0,
+        timeout: int | None = None,
+        mode: WORKER_WAIT_MODE = "at least",
+    ):
         info = await self.scheduler.identity()
         self._scheduler_identity = SchedulerInfo(info)
         if timeout:
             deadline = time() + parse_timedelta(timeout)
         else:
             deadline = None
 
-        def running_workers(info):
+        def running_workers(info, status_list=[Status.running]):
             return len(
                 [
                     ws
                     for ws in info["workers"].values()
-                    if ws["status"] == Status.running.name
+                    if ws["status"] in [s.name for s in status_list]
                 ]
             )
 
-        while n_workers and running_workers(info) < n_workers:
+        if mode == "at least":
+
+            def stop_condition(n_workers, info):
+                return running_workers(info) >= n_workers
+
+        elif mode == "exactly":
+
+            def stop_condition(n_workers, info):
+                return (
+                    running_workers(info, status_list=[Status.running, Status.paused])
+                    == n_workers
+                )
+
+        elif mode == "at most":
+
+            def stop_condition(n_workers, info):
+                return (
+                    running_workers(info, status_list=[Status.running, Status.paused])
+                    <= n_workers
+                )
+
+        else:
+            raise NotImplementedError(f"{mode} is not handled.")
+
+        while not stop_condition(n_workers, info):
```

Comment:
How about something like this, to avoid defining effectively the same code three times?

```python
from operator import eq, le, ge

try:
    op = {"at least": ge, "exactly": eq, "at most": le}[mode]
except KeyError:
    raise NotImplementedError(f"{mode} is not handled.")

while not op(running_workers(info, status_list[...]), n_workers):
   ...
```

Reply:
Yeah that's a cleaner approach. The `running_workers` calls are slightly different in a couple of them, do you have thoughts on how to handle that?

Reply:
Ah, sorry, I completely missed this go by.

```python
try:
    op, required_status = {"at least": (lt, xxx), "exactly": (ne, yyy), "at most": (gt, zzz)}[mode]
except KeyError:
    raise NotImplementedError(f"{mode} is not handled.")

while op(running_workers(info, status_list=required_status), n_workers):
   ...
```